enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,366 @@
+  `/usr/lib/microshift/components.d/10-storage-topolvm/` yamls, including crds.
+
+* The rebase script should be updated to extract those core components
+  as kustomize yamls instead of embedded assets.
+
+* Incompatible component rpms should be impossible to install together.
+  i.e. `microshift-networking-bridge` plus `microshift-networking-ovnk`,
+  this can be handled with Conflict spec clauses. (This is just an example,
+  since microshift-networking-bridge would not be necessary as it's the
+  default for cri-o)
+
+* Systemd services of the then optional networking components (like `microshift-ovs-init`)
+  should still start before the `microshift` systemd service, but `microshift` cannot
+  depend on `microshift-ovs-init` anymore. This can be acomplished by using the
+  Before=microshift clause in `microshift-ovs-init.service` and enabling this service
+  during the `microshift-networking-ovnk` sub-package install.
+
+### Workflow Description
+
+#### Deploying
+
+The workflow for deployment is already described in the `Proposal details` section.
+
+#### Upgrading
+
+Upgrading optional components would behave in the same way as upgrading
+components where the assets are embedded into the MicroShift binary.
+
+There is an open question about how to handle migration of CRD based
+content between versions. This problem is not handled or explored in MicroShift
+yet. Could this be handled by one-shot pods, the one-shot pods could be defined to
+be triggered only if an specific type of object/version existed, another
+alternative to this would be to use upgrade-binaries optional in components
+which are ran when necessary (similar to how CNIs work).
+
+#### Configuring
+
+Configuration of the optional components is done by package selection
+when the images are created.
+
+Specific component configuration could be handled as yaml files in
+`/etc/microshift/components.d/` . i.e.:
+
+```
+/etc/microshift/component.d/010-storage-topolvm/config.yaml:
+a:
+  b: 1234
+  c: 5677
+
+y: "abcd"
+```
+
+Would be rendered in the templates with:
+
+```
+{{ .config.a.b || 'default' }} {{ .config.a.c }} {{ config.y }}
+```
+
+Additionally more general MicroShift configuration and system details are passed
+down as variables to the templates:
+
+* `{{ .NodeName }}`
+* `{{ .NodeIP }}`
+* `{{ .ClusterCIDR }}`
+* `{{ .ServiceCIDR }}`
+* `{{ .ClusterDNS }}`
+* `{{ .BaseDomain }}`
+
+#### Deploying Applications
+
+Applications are deployed as usual, we would add support for multiple
+directories in the user component manifests as described in the 2nd goal.
+
+### API Extensions
+
+None.
+
+### Risks and Mitigations
+
+The component combinations would need additional testing handled by
+an increased testing matrix. Considering networking/storage to be independent
+elements, probably an additional set of tests which cover "microshift-core"
+alone would be enough. When running those tests we need to disable PV and
+NetworkPolicy related testing.
```

Comment:
I see the need for testing without the product CNI driver (at least upstream). I'm not sure I see the need for tests without the CSI driver installed. Do we think that leaving out CSI is going to cause some other core component to behave differently?

Reply:
You are right, no core components depend on CSI, so it's probably unnecessary to test without CSI, as you say.